validator/validator.py

```diff
@@ -0,0 +1,362 @@
+            "result": {
+                "summary": self.summary(),
+                "status": "OK",
+                "schema_url": self.schema_url,
+            }
+        }
+
+
+class ValidationRefOK(ValidationResult):
+    status = True
+
+    def __init__(self, kind, filename, ref, schema_url):
+        self.kind = kind
+        self.filename = filename
+        self.schema_url = schema_url
+        self.ref = ref
+
+    def dump(self):
+        return {
+            "filename": self.filename,
+            "ref": self.ref,
+            "kind": self.kind.value,
+            "result": {
+                "summary": self.summary(),
+                "status": "OK",
+                "schema_url": self.schema_url,
+                "ref": self.ref,
+            }
+        }
+
+
+class ValidationError(ValidationResult):
+    status = False
+
+    def __init__(self, kind, filename, reason, error, **kwargs):
+        self.kind = kind
+        self.filename = filename
+        self.reason = reason
+        self.error = error
+        self.kwargs = kwargs
+
+    def dump(self):
+        result = {
+            "summary": self.summary(),
+            "status": "ERROR",
+            "reason": self.reason,
+            "error": self.error.__str__()
+        }
+
+        result.update(self.kwargs)
+
+        return {
+            "filename": self.filename,
+            "kind": self.kind.value,
+            "result": result
+        }
+
+    def error_info(self):
+        if self.error.message:
+            msg = "{}\n{}".format(self.reason, self.error.message)
+        else:
+            msg = self.reason
+
+        return msg
+
+
+def get_handlers(schemas_bundle):
+    """
+    Generates a dictionary which will be used as an the `handlers` argument for
+    jsonschema.RefResolver.
+
+    `handlers` is a mapping from URI schemes to functions that should be used
+    to retrieve them.
+
+    In this case we are overloading the empty string scheme, which will be the
+    scheme detected for absolute or relative file paths.
+    """
+    return {
+        '': lambda s: schemas_bundle[s]
+    }
+
+
+def validate_schema(schemas_bundle, filename, schema_data):
+    kind = ValidatedFileKind.SCHEMA
+
+    logging.info('validating schema: {}'.format(filename))
+
+    try:
+        meta_schema_url = schema_data[u'$schema']
+    except KeyError as e:
+        return ValidationError(kind, filename, "MISSING_SCHEMA_URL", e)
+
+    if meta_schema_url in schemas_bundle:
+        meta_schema = schemas_bundle[meta_schema_url]
+    else:
+        meta_schema = fetch_schema(meta_schema_url)
+        schemas_bundle[meta_schema_url] = meta_schema
+
+    resolver = jsonschema.RefResolver(
+        filename,
+        schema_data,
+        handlers=get_handlers(schemas_bundle)
+    )
+
+    jsonschema.Draft4Validator.check_schema(schema_data)
+    validator = jsonschema.Draft4Validator(meta_schema, resolver=resolver)
+    validator.validate(schema_data)
```

Comment:
I'm super confused. We do this twice and the first time isn't in the `try/except`? This doesn't look intentional.

Reply:
It's not, I again forgot to clean it up... thanks for the eagle eye!